include/quicly/issued_cid.h

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2020 Fastly, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#ifndef issued_cid_h
+#define issued_cid_h
+
+#include "quicly/cid.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct st_quicly_issued_cid_set_t quicly_issued_cid_set_t;
+typedef struct st_quicly_issued_cid_t quicly_issued_cid_t;
+
+/**
+ * a callback function to generate a new connection ID in `cid`
+ *
+ * @return 0 if successfully generated, non-zero if unable to generate a CID (e.g. limit reached)
+ */
+typedef int (*quicly_issued_cid_generator_t)(quicly_conn_t *conn, quicly_issued_cid_t *cid);
+
+enum en_quicly_issued_cid_state_t {
+    /**
+     * this entry is free for use
+     */
+    QUICLY_ISSUED_CID_STATE_IDLE,
+    /**
+     * this entry is to be sent at the next round of send operation
+     */
+    QUICLY_ISSUED_CID_STATE_PENDING,
+    /**
+     * this entry has been sent and is waiting for ACK (or to be deemed lost)
+     */
+    QUICLY_ISSUED_CID_STATE_INFLIGHT,
+    /**
+     * this CID has been delivered to the peer (ACKed) and in use
+     */
+    QUICLY_ISSUED_CID_STATE_DELIVERED,
+};
+
+/**
+ * records information for sending NEW_CONNECTION_ID frame
+ */
+struct st_quicly_issued_cid_t {
+    enum en_quicly_issued_cid_state_t state;
+    uint64_t sequence;
+    quicly_cid_t cid;
+    uint8_t stateless_reset_token[QUICLY_STATELESS_RESET_TOKEN_LEN];
+};
+
+/**
+ * manages a list of connection IDs we issue to the peer
+ */
+struct st_quicly_issued_cid_set_t {
+    /**
+     * storage to retain issued CIDs
```

Comment:
👍 Actually, after another thought, I've renamed `_capacity` to `_size`. `_capacity` in C++ vector means the physical limit of the array, so it would have been confusion to people familiar with C++.

Reply:
Makes perfect sense. Actually, we use the two terms the same way (see `quicly_sendbuf_t::vecs` for example).